lib/ansible/template/__init__.py

```diff
@@ -152,10 +157,186 @@ def _count_newlines_from_end(in_str):
         while in_str[j] == '\n':
             j -= 1
         return i - 1 - j
+    except TypeError:
+        return 0
     except IndexError:
         # Uncommon cases: zero length string and string containing only newlines
         return i
 
+
+def ansible_j2_concat(invals):
+    '''
+    Jinja's native concat is just a list join. The list comes from result of
+    templating all the nodes. If the return type is expected to be
+    native, the list is artificial and we should return just the first
+    element.
+    '''
+    invals = [x for x in invals]
+    if isinstance(invals, list):
+        if len(invals) == 1:
+            # break out the single value
+            invals = invals[0]
+        elif len(invals) > 1:
+            # cast to unicode and join
+            invals = u''.join([u'%s' % x for x in invals])
+    return invals
```

Comment:
Profiling and jctanner's version is faster.  So the only things I can see to do differently:

* remove "if isinstance(invals, list):".  after the list comprehension, invals should always be a list.
* u'%s' % x for x in invals note that this isn't 100% safe (for instance b'café' will raise a unicode error) but I assume you took it from some part of jinja2 code so we aren't making a new assumption then.